### PR TITLE
Feature: cascading deletion of resources Fixes issue #333

### PR DESCRIPTION
**Pull request by bencurrerburgess**
This addresses issue #333 
This PR adds two flags to delete commands for Applications and Functions objects:

- **recursive (r)**:  Will retrieve all sub-objects and attempt to delete them from the bottom up before deleting the primary resource (user is prompted to continue after all of the associated resources are listed)
- **force (f)**: Will silence the prompt from a recursive delete and just deletes all resources without further user input (this only has an effect when used in combination with the recursive flag)

### example cascading
command:
```
fn delete fn <app-name> <fn-name> -r
```
will result in this prompt:
```
You are about to delete the following resources:
   Functions:    1
   Triggers:     <number of associated triggers>
This operation cannot be undone
Do you wish to proceed? y/n: 
```
if the user replies 'y' then:
```
Trigger  <trigger-name>  deleted
Function  <fn-name>  deleted
```
if the user replies 'n' then nothing is deleted:
```
Cancelling delete
```

### example cascading & forced
command:
```
fn delete app -f -r <app-name> 
```
results in:
```
Trigger  <trigger-name>  deleted
Function  <fn-name>  deleted
App <app-nam> deleted
```

### CODE REVIEW

**Comment by zootalures** [comment]
Would prefer -r over -c as it’s consistent with  (e.g rm)

**Comment by zootalures** [comment]
LGTM but could you write a couple of e2e tests (e.g. in cli_crud_test.go) ?  thanks

**Comment by bencurrerburgess** [comment]
> LGTM but could you write a couple of e2e tests (e.g. in cli_crud_test.go) ? thanks

Added this in now, though it would seem that the tests on master are broken already, but the test I have written passes. 👍

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ data/
 .idea/
 **/*.iml
 **/.DS_Store
+/vendor
```
Comment (hatchan) [inline .gitignore]: The git repository contains a checked in `vendor` directory. Are you sure this needs to be ignored?

File: common/common.go [hunk 2]
```diff
@@ -595,3 +599,142 @@ func GetFuncYamlVersion(oldFF map[string]interface{}) int {
 	}
 	return 1
 }
+
+// UserConfirmedMultiResourceDeleteion will prompt the user for confirmation to delete all the the resources
+func UserConfirmedMultiResourceDeleteion(aps []*modelsv2.App, fns []*modelsv2.Fn, trs []*modelsv2.Trigger) bool {
```
Comment (hatchan) [inline common/common.go]: `Deleteion` should be spelled `Deletion`

File: common/common.go [hunk 3]
```diff
@@ -595,3 +599,142 @@ func GetFuncYamlVersion(oldFF map[string]interface{}) int {
...
+		return false
+	} else {
+		fmt.Println("Unrecognised input, should be y/n")
+		fmt.Println("Cancelling delete")
+		return false
+	}
+	return true
+}
+
+// ListFnsAndTriggersInApp lists all the fucntions associated with an app and all the triggers associated with each of those functions
```
Comment (hatchan) [inline common/common.go]: `fucntions` should be `functions`

File: common/common.go [hunk 4]
```diff
@@ -595,3 +599,142 @@ func GetFuncYamlVersion(oldFF map[string]interface{}) int {
...
+	fmt.Printf("Do you wish to proceed? Y/N: ")
+	reader := bufio.NewReader(os.Stdin)
+	input, _ := reader.ReadString('\n')
+	if strings.ToLower(input) == "y\n" {
+		return true
+	} else if strings.ToLower(input) == "n\n" {
+		fmt.Println("Cancelling delete")
+		return false
+	} else {
+		fmt.Println("Unrecognised input, should be y/n")
```
Comment (hatchan) [inline common/common.go]: The `y/n` should probably match the casing of the first query, so `Y/N`

File: common/common.go [hunk 5]
```diff
@@ -595,3 +599,142 @@ func GetFuncYamlVersion(oldFF map[string]interface{}) int {
...
+		fmt.Println("   Functions:   ", fnsLen)
+	}
+	if trsLen > 0 {
+		fmt.Println("   Triggers:    ", trsLen)
+	}
+	fmt.Println("This operation cannot be undone")
+	fmt.Printf("Do you wish to proceed? Y/N: ")
+	reader := bufio.NewReader(os.Stdin)
+	input, _ := reader.ReadString('\n')
+	if strings.ToLower(input) == "y\n" {
```
Comment (hatchan) [inline common/common.go]: You might want to consider running this through `strings.TrimSpace` to get rid of any spaces and newlines (https://golang.org/pkg/strings/#TrimSpace)

**Pull request merged**